Nodsoft.WowsReplaysUnpack/Services/ReplayUnpackerService.cs

```diff
@@ -100,14 +106,13 @@ public UnpackedReplay Unpack(Stream stream, ReplayUnpackerOptions? options = nul
 		// Decrypted stream not used anymore
 		decryptedStream.Dispose();
 
-
 		Version gameclientVersion = Version.Parse(arenaInfo.ClientVersionFromExe.Replace(',', '.'));
-		
+		_semaphore.Wait(_semephoreTimeOut);
```

Comment:
I saw this was added to "prevent a deadlock" but i'm curious what resource was actually deadlocked?
And this should definitely be a configurable option.

Reply:
I believe the deadlock still exists in this pull request when there are many threads like 12 or even more. It is also likely to happen in test. The reason behind is still unknown, I believe some functions are still holding the semephore in rare occasions.
The true root cause was that the implementation didn't consider thread safety in the first place. A rewrite may be required for this to work properly. However, it can be time consuming.

Reply:
@HenryQuan do you know what is gets deadlocked though? Does it only happen when files are read from the filesystem or also when it's coming from memory? Or is it purely because it's not async? I'm fairly certain i had it async at the start but then changed it back because it made no sense somewhere...

Reply:
The issue without semephores was due to the decryption. Please see, https://github.com/Nodsoft/WoWS-ReplaysUnpack/pull/29#issue-1921067190. I don't understand why `BinaryReader` isn't thread safe. It seems that other threads are changing its pointer also as it is like a singleton. With semephores, in a heavy multithread condition, there may be some deadlocks. Not sure what is blocking it because there shouldn't be any locks. We probably to investigate further on this. I am not very familar with C# semephore also.

Reply:
Ah, so initially the issue wasn't deadlocks but some weird errors with the BinaryReader and now with the added Semaphore there's the potential for deadlocks?

Also the issue you mentioned is about the new file system file store right?

So am i right in assuming the previous issue and the deadlocks only happen with the file system store and not when it's coming from memory?

Reply:
That's correct I believe. Maybe, it is a lock coming from the file system. I have not tested with the memory approach yet.

Reply:
That would make more sense yeah. I wouldn't even know where to begin if it's not caused by the file system...
I just had a quick look at the code cause i couldn't quite remember but i think the issue could be that the definitions are loaded on-demand and then cached for future access.
This wasn't an issue when i first implemented it with the embedded resources i guess, but when you open them from the FS there's a potential for file locks or whatever. So maybe the filesystem store needs to cache all definitions on initialization